maven/PomGenerator.kt

```diff
@@ -0,0 +1,212 @@
+        val transformer = TransformerFactory.newInstance().newTransformer()
+        transformer.setOutputProperty(OutputKeys.INDENT, "yes")
+        transformer.setOutputProperty("{http://xml.apache.org/xslt}indent-amount", "2")
+        transformer.transform(DOMSource(pom), StreamResult(outputFile))
+    }
+
+    override fun call() {
+        val version = versionFile.readText()
+        val workspace_refs = Json.parse(workspaceRefsFile.readText()).asObject()
+
+        // Create an XML document for constructing the POM
+        val pom = DocumentBuilderFactory.newInstance().newDocumentBuilder().newDocument()
+
+        // root element describing the project
+        val rootElement = pom.createElement("project")
+        rootElement.setAttribute("xmlns", "http://maven.apache.org/POM/4.0.0")
+        rootElement.setAttribute("xmlns:xsi", "http://www.w3.org/2001/XMLSchema-instance")
+        rootElement.setAttribute("xsi:schemaLocation", "http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd")
+        pom.appendChild(rootElement)
+
+        // version of the Maven POM format
+        val modelVersion = pom.createElement("modelVersion")
+        modelVersion.appendChild(pom.createTextNode("4.0.0"))
+        rootElement.appendChild(modelVersion)
+
+        // project name
+        val name = pom.createElement("name")
+        name.appendChild(pom.createTextNode(projectName))
+        rootElement.appendChild(name)
+
+        // project description
+        val description = pom.createElement("description")
+        description.appendChild(pom.createTextNode(projectDescription))
+        rootElement.appendChild(description)
+
+        // project URL
+        val url = pom.createElement("url")
+        url.appendChild(pom.createTextNode(projectUrl))
+        rootElement.appendChild(url)
+
+        // licenses
+        rootElement.appendChild(licenses(pom))
+
+        // source control management information
+        rootElement.appendChild(scm(pom, version))
+
+        // group id of the library
+        val groupIdElem = pom.createElement("groupId")
+        groupIdElem.appendChild(pom.createTextNode(targetGroupId))
+        rootElement.appendChild(groupIdElem)
+
+        // artifact id of the library
+        val artifactIdElem = pom.createElement("artifactId")
+        artifactIdElem.appendChild(pom.createTextNode(targetArtifactId))
+        rootElement.appendChild(artifactIdElem)
+
+        // version of the library
+        val versionElem = pom.createElement("version")
+        versionElem.appendChild(pom.createTextNode(version))
+        rootElement.appendChild(versionElem)
+
+        // add dependency information
+        rootElement.appendChild(dependencies(pom, version, workspace_refs))
+
+        // write the final result
+        outputDocumentToFile(pom)
+    }
+}
+
+fun main(args: Array<String>): Unit = exitProcess(CommandLine(PomGenerator()).execute(*args))
```

Comment:
Here too. Can we remove the `call` method and do the work in `main` instead?

Reply:
@lolski [this](https://github.com/graknlabs/bazel-distribution/pull/293#discussion_r601386228) applies here too